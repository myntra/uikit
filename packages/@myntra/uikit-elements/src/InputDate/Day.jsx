import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import styles from './InputDate.css'
import { classnames } from '@myntra/uikit-utils'

export default class Day extends PureComponent {
  static propTypes = {
    date: PropTypes.instanceOf(Date),
    isEmpty: PropTypes.bool,
    isStart: PropTypes.bool,
    isEnd: PropTypes.bool,
    isDisabled: PropTypes.bool,
    isFocused: PropTypes.bool,
    isSelected: PropTypes.bool.isRequired,
    onSelect: PropTypes.func.isRequired,
    onFocus: PropTypes.func
  }

  onFocus(event) {
    if (!this.props.isFocused && this.props.onFocus) {
      this.props.onFocus(event, this.props.isDisabled ? null : this.props.date)
    }
  }

  handleMouseDown = event => {
    if (!this.props.isEmpty && !this.props.isDisabled) {
      event.preventDefault()
      event.stopPropagation()

      this.props.onSelect(event, this.props.date)
    }
  }

  handleMouseEnter = event => {
    this.onFocus(event)
  }

  handleMouseMove = event => {
    this.onFocus(event)
  }

  handleTouchStart = () => {
    this.dragging = false
  }

  handleTouchMove = () => {
    this.dragging = true
  }

  handleTouchEnd = event => {
    if (this.dragging) return

    this.handleMouseDown(event)
  }

  render() {
    const { date, isDisabled, isSelected, isFocused, isStart, isEmpty, isEnd } = this.props
    const isToday = date && date.toDateString() === new Date().toDateString()

    return (
      <td
        className={classnames('gridDay', {
          gridDayDisabled: isDisabled,
          gridDayFocused: isFocused,
          gridDaySelected: isSelected,
          gridDaySelectedStart: isSelected && isStart,
          gridDaySelectedEnd: isSelected && isEnd,
          gridDayEmpty: isEmpty,
          gridToday: isToday
        }).use(styles)}
        title={date && date.toDateString()}
        onClick={this.handleMouseDown}
        onMouseEnter={this.handleMouseEnter}
        onTouchStart={this.handleTouchStart}
        onTouchMove={this.handleTouchMove}
        onTouchEnd={this.handleTouchEnd}
      >
        {!isEmpty && date ? date.getDate() : ''}
      </td>
    )
  }
}
