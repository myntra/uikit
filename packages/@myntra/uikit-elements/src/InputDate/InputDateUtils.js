import PropTypes from 'prop-types'
import { parse as baseParse, format as baseFormat } from 'date-fns'

export const RE_ISO_DATE = /^(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+)|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d)|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d)$/

export const DateType = PropTypes.oneOfType([PropTypes.instanceOf(Date), PropTypes.string])

export const DateRangeType = PropTypes.oneOfType([
  PropTypes.shape({
    from: PropTypes.string,
    to: PropTypes.string
  }),
  PropTypes.shape({
    from: PropTypes.instanceOf(Date),
    to: PropTypes.instanceOf(Date)
  })
])

const padStartWithZeros = (value, digits = 2) => `${value}`.padStart(digits, '0')

/**
 * Create date instance.
 *
 * @export
 * @param {number} year
 * @param {number} month
 * @param {number} date
 * @returns {Date}
 */
export function UTCDate(year, month, date) {
  return new Date(
    `${padStartWithZeros(year, 4)}-${padStartWithZeros(month + 1)}-${padStartWithZeros(date)}T00:00:00.000Z`
  )
}

/**
 * Drop time component from Date.
 *
 * @export
 * @param {Date} date
 * @returns {Date}
 */
export function onlyDate(date) {
  if (typeof date.toDate === 'function') {
    date = date.toDate() // dayjs interop.
  }

  return UTCDate(date.getFullYear(), date.getMonth(), date.getDate())
}

export function parse(str, format) {
  if (str instanceof Date) return str
  if (typeof str !== 'string') return null
  if (!format) throw new Error(`'format' is required.`)
  if (RE_ISO_DATE.test(str)) return new Date(str)
  if (str.length !== format.length) throw new Error(`Date parse exception: unequal length`)

  const date = baseParse(str, format, new Date())

  if (date instanceof Date) {
    return onlyDate(date)
  }

  throw new Error(`Date parse exception: '${str}' does not match '${format}'`)
}

export function format(value, pattern) {
  if (!value) return value
  if (!pattern) return value
  if (value instanceof Date) return baseFormat(value, pattern)
  if (typeof value === 'string') return format(parse(value, pattern), pattern)

  const result = {}
  if (value.from) result.from = format(value.from, pattern)
  if (value.to) result.to = format(value.to, pattern)
  return result
}

export function isDateEqual(a, b) {
  if (!(a instanceof Date)) return false
  if (!(b instanceof Date)) return false
  if (a.getTime() === b.getTime()) return true

  return a.getFullYear() === b.getFullYear() && a.getMonth() === b.getMonth() && a.getDate() === b.getDate()
}
