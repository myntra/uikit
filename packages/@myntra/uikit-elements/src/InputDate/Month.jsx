import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import Day from './Day'
import { UTCDate } from './utils'

import styles from './InputDate.css'

function isSelected(d, selected) {
  if (selected.from && selected.to) {
    return selected.from <= d && d <= selected.to
  }
  if (selected.from) return selected.from <= d
  if (selected.to) return d <= selected.to
  return false
}

function isEqual(a, b) {
  if (!a || !b) return false

  return a.getFullYear() === b.getFullYear() && a.getMonth() === b.getMonth() && a.getDate() === b.getDate()
}

function isSelectedLoose(d, selected, isFirst) {
  if (isFirst && isEqual(d, selected.from)) return false
  if (!isFirst && isEqual(d, selected.to)) return false

  return isSelected(d, selected)
}

function chunk(items, size) {
  const length = Math.ceil(items.length / size)

  // eslint-disable-next-line
  return Array.apply(null, { length }).map((_, i) => items.slice(i * size, i * size + size))
}

const dayHeaders = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday']

export default class Month extends PureComponent {
  static propTypes = {
    month: PropTypes.number.isRequired, // zero indexed month. 0 - Jan, 1 - Feb ...
    year: PropTypes.number.isRequired,
    selected: PropTypes.shape({
      from: PropTypes.instanceOf(Date),
      to: PropTypes.instanceOf(Date)
    }).isRequired,
    focused: PropTypes.instanceOf(Date),
    disabled: PropTypes.oneOfType([
      PropTypes.shape({
        from: PropTypes.instanceOf(Date).isRequired,
        to: PropTypes.instanceOf(Date).isRequired
      }),
      PropTypes.arrayOf(PropTypes.instanceOf(Date))
    ]),
    onDateFocus: PropTypes.func,
    onSelect: PropTypes.func.isRequired,
    minDate: PropTypes.instanceOf(Date),
    maxDate: PropTypes.instanceOf(Date)
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.disabled !== this.props.disabled) {
      this._disabled = null
    }
  }

  get selected() {
    return this.props.selected || {}
  }

  get disabled() {
    if (this._disabled) return this._disabled

    let dates = []
    if (Array.isArray(this.props.disabled)) {
      dates = this.props.disabled
    }
    if (this.props.disabled) {
      const it = new Date(this.props.disabled.from)
      while (it < this.props.disabled.to) {
        dates.push(it)

        it.setDate(it.getDate() + 1)
      }
    }

    this._disabled = dates.reduce((acc, it) => ({ ...acc, [it.toDateString()]: true }), {})

    return this._disabled
  }

  handleDaySelection = (event, date) => {
    this.props.onSelect(date)
  }

  handleDayFocus = (event, date) => {
    if (this.props.onDateFocus) this.props.onDateFocus(date)
  }

  render() {
    const { month, year, minDate, maxDate } = this.props
    const selected = this.selected
    const disabled = this.disabled
    const firstOfMonth = UTCDate(year, month, 1)
    const firstDayOfWeek = firstOfMonth.getDay()
    const days = []
    const isFirstSelected = isSelectedLoose(firstOfMonth, selected, true)

    for (let i = 0; i < firstDayOfWeek; i += 1) {
      days.push(<Day isEmpty isSelected={isFirstSelected} onSelect={() => {}} onFocus={this.handleDayFocus} key={i} />)
    }
    const it = new Date(firstOfMonth)
    let lastDateOfMonth = 1
    while (it.getMonth() === firstOfMonth.getMonth()) {
      lastDateOfMonth = it.getDate()
      const isOutOfRange = (minDate && minDate > it) || (maxDate && maxDate < it)
      days.push(
        <Day
          key={it}
          date={new Date(it)}
          isDisabled={disabled[it.toDateString()] || isOutOfRange}
          isFocused={this.props.focused && it.toDateString() === this.props.focused.toDateString()}
          isStart={isEqual(it, selected.from)}
          isEnd={isEqual(it, selected.to)}
          isSelected={isSelected(it, selected)}
          onSelect={this.handleDaySelection}
          onFocus={this.handleDayFocus}
        />
      )
      it.setDate(it.getDate() + 1)
    }

    const lastOfMonth = new Date(year, month, lastDateOfMonth, 0, 0, 0, 0)
    const lastDayOfMonth = lastOfMonth.getDay()
    const isLastSelected = isSelectedLoose(lastOfMonth, selected)

    for (let i = lastDayOfMonth + 1; i < 7; i += 1) {
      days.push(<Day isEmpty isSelected={isLastSelected} onSelect={() => {}} onFocus={this.handleDayFocus} key={i} />)
    }

    const weeks = chunk(days, 7).map((week, i) => (
      <tr className={styles.gridWeek} key={i}>
        {week}
      </tr>
    ))

    return (
      <table className={styles.gridMonth}>
        <thead className={styles.gridWeekHeader}>
          <tr>
            {dayHeaders.map(h => (
              <th className={styles.gridDayHeader} title={h} key={h}>
                {h.charAt(0)}
              </th>
            ))}
          </tr>
        </thead>
        <tbody>{weeks}</tbody>
      </table>
    )
  }
}
