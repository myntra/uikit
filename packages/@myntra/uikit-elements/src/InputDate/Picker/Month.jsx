import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { memoize, classnames } from '@myntra/uikit-utils'

import Day from './Day'

import styles from './Month.module.css'
import { UTCDate } from '../InputDateUtils'
import dayJS from 'dayjs'

const DAYS_OF_WEEK = 'Sunday,Monday,Tuesday,Wednesday,Thursday,Friday,Saturday'.split(',')
const CLASS = {
  month: classnames('month').use(styles),
  header: classnames('header').use(styles),
  day: classnames('day-header').use(styles),
  days: classnames('days').use(styles)
}

/**
 * @since 0.0.0
 * @status REVIEWING
 * @example
 * <InputDatePicker.Month year={2016} month={2} />
 */
export default class Month extends PureComponent {
  static propTypes = {
    /** @private */
    className: PropTypes.string,
    /** @private */
    children: PropTypes.any,
    year: PropTypes.number.isRequired,
    /** zero indexed month. 0 - Jan, 1 - Feb ... */
    month: PropTypes.number.isRequired,
    focused: PropTypes.number,
    selected: PropTypes.shape({ from: PropTypes.number, to: PropTypes.number }),
    disabled: PropTypes.arrayOf(PropTypes.shape({ from: PropTypes.number, to: PropTypes.number })),
    /**
     * @function
     * @param {Date} value
     */
    onDateFocus: PropTypes.func,
    /**
     * @function
     * @param {Date} value
     */
    onDateSelect: PropTypes.func
  }

  static defaultProps = {
    selected: {},
    disabled: []
  }

  makeDayValidator(ranges) {
    const daysInCurrentMonth =
      dayJS(UTCDate(this.props.year, this.props.month, 1))
        .endOf('month')
        .toDate()
        .getUTCDate() + 1
    const isValid = Array.apply(null, { length: daysInCurrentMonth + 1 }).map(Boolean)

    ranges.forEach(({ from, to }) => {
      for (let i = from; i <= to; ++i) isValid[Math.min(i, daysInCurrentMonth)] = true
    })

    return day => isValid[Math.min(day, daysInCurrentMonth)]
  }

  dayDisabledValidator = memoize(args => this.makeDayValidator(args))

  daySelectedValidator(first, last) {
    const hasSelection = this.props.selected && ('from' in this.props.selected || 'to' in this.props.selected)
    const from =
      hasSelection && 'from' in this.props.selected ? this.props.selected.from : this.props.focused || this.lastFocus
    const to =
      hasSelection && 'to' in this.props.selected ? this.props.selected.to : this.props.focused || this.lastFocus
    const low = Math.min(from || 0, to)
    const high = Math.max(from, to || last.getUTCDate() + 1)

    if (this.props.focused) this.lastFocus = this.props.focused

    return date => hasSelection && low <= date && date <= high
  }

  handleDateSelection = date => {
    if (this.props.onDateSelect) this.props.onDateSelect(date.getUTCDate(), date)
  }

  handleDateFocus = date => {
    if (this.props.onDateFocus) this.props.onDateFocus(date ? date.getUTCDate() : null, date)
  }

  render() {
    const { month, year, focused } = this.props
    const isDayDisabled = this.dayDisabledValidator(this.props.disabled)
    const dateOnFirstOfMonth = UTCDate(year, month, 1)
    const dateOnLastOfMonth = dayJS(dateOnFirstOfMonth)
      .endOf('month')
      .toDate()
    const isDaySelected = this.daySelectedValidator(dateOnFirstOfMonth, dateOnLastOfMonth)
    const days = []

    for (let i = 0; i < dateOnFirstOfMonth.getDay(); i += 1)
      days.push(
        <Day
          key={'pre-' + i}
          selected={isDaySelected(0 /* selection starts in previous month. */)}
          disabled={isDayDisabled(0)}
          onFocus={this.handleDateFocus}
        />
      )
    for (let day = 1; day <= dateOnLastOfMonth.getUTCDate(); day += 1)
      days.push(
        <Day
          key={'day-' + year + '-' + month + '-' + day}
          day={day}
          year={year}
          month={month}
          focused={day === focused}
          isSelectionStart={!isDaySelected(day - 1) && isDaySelected(day)}
          isSelectionEnd={isDaySelected(day) && !isDaySelected(day + 1)}
          disabled={isDayDisabled(day)}
          selected={isDaySelected(day)}
          onSelect={this.handleDateSelection}
          onFocus={this.handleDateFocus}
        />
      )

    for (let i = dateOnLastOfMonth.getDay() + 1; i < 7; i += 1)
      days.push(
        <Day
          key={'post' + i}
          selected={isDaySelected(dateOnLastOfMonth.getUTCDate() + 1 /* selection ends in next month */)}
          disabled={isDayDisabled(dateOnLastOfMonth.getUTCDate() + 1)}
          onFocus={this.handleDateFocus}
        />
      )

    return (
      <div className={classnames(this.props.className, CLASS.month)}>
        {this.props.children}
        <div className={CLASS.header}>
          {DAYS_OF_WEEK.map(day => (
            <div className={CLASS.day} title={day} key={day}>
              {day.charAt(0)}
            </div>
          ))}
        </div>
        <div className={CLASS.days}>{days}</div>
      </div>
    )
  }
}
