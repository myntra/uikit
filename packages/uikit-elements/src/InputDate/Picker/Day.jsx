import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import classnames from './Day.module.css'
import { UTCDate } from '../InputDateUtils'

/**
 * @since 0.0.0
 * @status REVIEWING
 * @example
 * <InputDatePicker.Day year={2018} month={0} day={1} />
 */
export default class Day extends PureComponent {
  static propTypes = {
    year: PropTypes.number,
    month: PropTypes.number,
    day: PropTypes.number,
    focused: PropTypes.bool,
    disabled: PropTypes.bool,
    selected: PropTypes.bool,
    isSelectionStart: PropTypes.bool,
    isSelectionEnd: PropTypes.bool,
    /**
     * @function
     * @param {Date} value
     */
    onSelect: PropTypes.func,
    /**
     * @function
     * @param {Date} value
     */
    onFocus: PropTypes.func,
    /** @private */
    isValidDate(props) {
      const names = ['year', 'month', 'day']

      if (names.every(name => typeof props[name] === 'number')) {
        if (Number.isNaN(UTCDate(props.year, props.month, props.day).getTime())) {
          throw new Error(`Invalid date`)
        }
      }
    }
  }

  get isEmpty() {
    return Number.isNaN(this.date.getTime())
  }

  get date() {
    return UTCDate(this.props.year, this.props.month, this.props.day)
  }

  handleFocus = event => {
    if (this.props.onFocus) {
      if (this.isEmpty || this.props.disabled) {
        this.props.onFocus(null)
      } else if (!this.props.focused) this.props.onFocus(this.date)
    }
  }

  handleBlur = event => {
    if (this.props.onFocus && this.props.focused) {
      this.props.onFocus(null)
    }
  }

  handleMouseDown = event => {
    if (this.props.disabled) {
      event.stopPropagation()
    } else if (this.props.onSelect && !this.isEmpty) {
      this.props.onSelect(this.date)
    }
  }

  handleMouseEnter = event => this.handleFocus(event)
  handleMouseLeave = event => this.handleBlur(event)

  handleTouchStart = () => {
    this.dragging = false
  }
  handleTouchMove = () => {
    this.dragging = true
  }
  handleTouchEnd = event => {
    if (this.dragging) return

    this.handleMouseDown(event)
  }

  render() {
    const { date, isEmpty } = this
    const { disabled, focused, selected, isSelectionEnd, isSelectionStart } = this.props
    const isToday = !isEmpty && date.toDateString() === new Date().toDateString()

    return (
      <div
        className={classnames('day', {
          focused,
          selected,
          start: selected && isSelectionStart,
          end: selected && isSelectionEnd,
          empty: isEmpty,
          today: isToday
        })}
        aria-hidden={isEmpty}
        disabled={disabled}
        title={isEmpty ? null : this.date.toDateString()}
        onMouseDown={this.handleMouseDown}
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.handleMouseLeave}
        onTouchStart={this.handleTouchStart}
        onTouchMove={this.handleTouchMove}
        onTouchEnd={this.handleTouchEnd}
      >
        {isEmpty ? 0 : date.getUTCDate()}
      </div>
    )
  }
}
