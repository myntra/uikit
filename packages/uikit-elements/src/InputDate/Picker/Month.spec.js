import React from 'react'
import { shallow, mount } from 'enzyme'
import Month from './Month'
import Day from './Day'
import { UTCDate } from '../InputDateUtils'

describe('render', () => {
  it('should render', () => {
    const wrapper = shallow(<Month year={2018} month={4} />)
    const days = wrapper.find(Day)

    expect(days).toHaveLength(35)
    expect(days.at(2).props().day).toBe(1)
    expect(wrapper).toMatchSnapshot()
  })

  it('should focus date', () => {
    const wrapper = shallow(<Month year={2018} month={4} focused={1} />)
    const days = wrapper.find(Day)

    expect(days).toHaveLength(35)
    expect(days.at(2).props().day).toBe(1)
    expect(days.at(2).props().focused).toBe(true)
  })

  it('should render selected date', () => {
    const wrapper = shallow(<Month year={2018} month={4} selected={{ from: 1, to: 5 }} />)
    const days = wrapper.find(Day)

    expect(days.at(1).props().selected).toBe(false) // empty
    expect(days.at(2).props().selected).toBe(true) // 1
    expect(days.at(7).props().selected).toBe(false) // 6
  })

  it('should render selected date spanning multiple months (start)', () => {
    const wrapper = shallow(<Month year={2018} month={4} selected={{ from: 0, to: 5 }} />)
    const days = wrapper.find(Day)

    expect(days.at(1).props().selected).toBe(true) // empty
    expect(days.at(2).props().selected).toBe(true) // 1
    expect(days.at(7).props().selected).toBe(false) // 6
  })

  it('should render selected date spanning multiple months (end)', () => {
    const wrapper = shallow(<Month year={2018} month={4} selected={{ from: 2, to: 32 }} />)
    const days = wrapper.find(Day)

    expect(days.at(1).props().selected).toBe(false) // empty
    expect(days.at(4).props().selected).toBe(true) // 3
    expect(days.at(34).props().selected).toBe(true) // 6
  })

  it('should render disabled date', () => {
    const wrapper = shallow(<Month year={2018} month={4} disabled={[{ from: 1, to: 5 }, { from: 10, to: 15 }]} />)
    const days = wrapper.find(Day)

    expect(days.at(1).props().disabled).toBe(false) // empty
    expect(days.at(2).props().disabled).toBe(true) // 1
    expect(days.at(7).props().disabled).toBe(false) // 6
    expect(days.at(11).props().disabled).toBe(true) // 10
    expect(days.at(17).props().disabled).toBe(false) // 16
  })

  it('should render children', () => {
    const wrapper = shallow(
      <Month year={2018} month={4}>
        <span className="test-child">Foo</span>
      </Month>
    )

    expect(wrapper.find('.test-child').text()).toBe('Foo')
  })
})

describe('interaction', () => {
  describe('focus', () => {
    it('should trigger focus', () => {
      const fn = jest.fn()
      const wrapper = mount(<Month year={2018} month={4} onDateFocus={fn} />)

      wrapper
        .find(Day)
        .at(2)
        .simulate('mouseenter')

      expect(fn).toHaveBeenCalledWith(1, UTCDate(2018, 4, 1))
    })

    it('should trigger focus on empty day', () => {
      const fn = jest.fn()
      const wrapper = mount(<Month year={2018} month={4} onDateFocus={fn} />)

      wrapper
        .find(Day)
        .at(0)
        .simulate('mouseenter')

      expect(fn).toHaveBeenCalledWith(null, null)
    })

    it('should work without onDateFocus', () => {
      const wrapper = mount(<Month year={2018} month={4} />)

      expect(() =>
        wrapper
          .find(Day)
          .at(2)
          .simulate('mouseenter')
      ).not.toThrow()
    })
  })

  describe('select', () => {
    it('should trigger select', () => {
      const fn = jest.fn()
      const wrapper = mount(<Month year={2018} month={4} onDateSelect={fn} />)

      wrapper
        .find(Day)
        .at(2)
        .simulate('mousedown')

      expect(fn).toHaveBeenCalledWith(1, UTCDate(2018, 4, 1))
    })

    it('should work without onDateSelect', () => {
      const wrapper = mount(<Month year={2018} month={4} />)

      expect(() =>
        wrapper
          .find(Day)
          .at(2)
          .simulate('mousedown')
      ).not.toThrow()
    })
  })
})
