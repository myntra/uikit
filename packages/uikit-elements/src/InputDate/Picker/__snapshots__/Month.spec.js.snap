// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should render 1`] = `
<div
  className="month"
>
  <div
    className="header"
  >
    <div
      className="day-header"
      key="Sunday"
      title="Sunday"
    >
      S
    </div>
    <div
      className="day-header"
      key="Monday"
      title="Monday"
    >
      M
    </div>
    <div
      className="day-header"
      key="Tuesday"
      title="Tuesday"
    >
      T
    </div>
    <div
      className="day-header"
      key="Wednesday"
      title="Wednesday"
    >
      W
    </div>
    <div
      className="day-header"
      key="Thursday"
      title="Thursday"
    >
      T
    </div>
    <div
      className="day-header"
      key="Friday"
      title="Friday"
    >
      F
    </div>
    <div
      className="day-header"
      key="Saturday"
      title="Saturday"
    >
      S
    </div>
  </div>
  <div
    className="days"
  >
    <Day
      disabled={false}
      key="pre-0"
      onFocus={[Function]}
      selected={false}
    />
    <Day
      disabled={false}
      key="pre-1"
      onFocus={[Function]}
      selected={false}
    />
    <Day
      day={1}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-1"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={2}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-2"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={3}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-3"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={4}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-4"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={5}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-5"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={6}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-6"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={7}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-7"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={8}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-8"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={9}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-9"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={10}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-10"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={11}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-11"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={12}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-12"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={13}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-13"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={14}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-14"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={15}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-15"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={16}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-16"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={17}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-17"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={18}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-18"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={19}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-19"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={20}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-20"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={21}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-21"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={22}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-22"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={23}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-23"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={24}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-24"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={25}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-25"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={26}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-26"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={27}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-27"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={28}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-28"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={29}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-29"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={30}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-30"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      day={31}
      disabled={false}
      focused={false}
      isSelectionEnd={false}
      isSelectionStart={false}
      key="day-2018-4-31"
      month={4}
      onFocus={[Function]}
      onSelect={[Function]}
      selected={false}
      year={2018}
    />
    <Day
      disabled={false}
      key="post5"
      onFocus={[Function]}
      selected={false}
    />
    <Day
      disabled={false}
      key="post6"
      onFocus={[Function]}
      selected={false}
    />
  </div>
</div>
`;
